docs/Threading.md

```diff
@@ -0,0 +1,167 @@
+The SGX-LKL threading implementation
+====================================
+
+SGX-LKL has four kinds of threads: ethreads, lthreads, Linux tasks, and pthreads.
+This document explains how they are implemented and how they relate.
+
+Ethreads are the lowest-level threads are ethreads.
+These are roughly analogous to VCPUs in conventional VMs.
+In SGX, each ethread corresponds to a thread-control structure (TCS), allocated by OpenEnclave.
+
+The lthreads layer provides a cooperative threading library on top of ethreads.
+Each ethread runs an instance of the lthread scheduler, which pulls lthreads from a multi-producer, multi-consumer queue (MPMCQ) to run.
+These threads run until they explicitly yield.
+
+The Linux kernel (LKL) manages its own task abstraction.
+Each Linux task is associated with an lthread.
+The LKL documentation refers to 'host threads', these are lthreads in our use.
+
+The userspace libc (musl) provides a POSIX threads (pthreads) implementation that sits atop Linux's tasks.
+Each pthread is run by the lthread scheduler but may be caused to block by LKL.
+
```

Comment:
considering that you refer to "lowest level" in the beginning a diagram of "levels" and where each of the 4 types exist would be very helpful.